test/lib/jdk/test/lib/serial/SerialObjectBuilder.java

```diff
@@ -0,0 +1,217 @@
+        return this;
+    }
+
+    public <T> SerialObjectBuilder addPrimitiveField(String name, Class<T> type, T value) {
+        if (!type.isPrimitive())
+            throw new IllegalArgumentException("Unexpected non-primitive field: " + type);
+        primFields.put(new NameAndType<>(name, type), value);
+        return this;
+    }
+
+    public <T> SerialObjectBuilder addField(String name, Class<T> type, T value) {
+        if (type.isPrimitive())
+            throw new IllegalArgumentException("Unexpected primitive field: " + type);
+        objectFields.put(new NameAndType<>(name, type), value);
+        return this;
+    }
+
+    private static int getPrimitiveSignature(Class<?> cl) {
+        if (cl == Integer.TYPE) return 'I';
+        else if (cl == Byte.TYPE) return 'B';
+        else if (cl == Long.TYPE) return 'J';
+        else if (cl == Float.TYPE) return 'F';
+        else if (cl == Double.TYPE) return 'D';
+        else if (cl == Short.TYPE) return 'S';
+        else if (cl == Character.TYPE) return 'C';
+        else if (cl == Boolean.TYPE) return 'Z';
+        else throw new InternalError();
+    }
+
+    private static void writeUTF(DataOutputStream out, String str) throws IOException {
+        int utflen = str.length(); // assume ASCII
+        assert utflen <= 0xFFFF;
+        out.writeShort(utflen);
+        out.writeBytes(str);
+    }
+
+    private void writePrimFieldsDesc(DataOutputStream out) throws IOException {
+        for (Map.Entry<NameAndType<?>, Object> entry : primFields.entrySet()) {
+            assert entry.getKey().type() != void.class;
+            out.writeByte(getPrimitiveSignature(entry.getKey().type())); // prim_typecode
+            out.writeUTF(entry.getKey().name());                         // fieldName
+        }
+    }
+
+    private void writePrimFieldsValues(DataOutputStream out) throws IOException {
+        for (Map.Entry<NameAndType<?>, Object> entry : primFields.entrySet()) {
+            Class<?> cl = entry.getKey().type();
+            Object value = entry.getValue();
+            if (cl == Integer.TYPE) out.writeInt((int) value);
+            else if (cl == Byte.TYPE) out.writeByte((byte) value);
+            else if (cl == Long.TYPE) out.writeLong((long) value);
+            else if (cl == Float.TYPE) out.writeFloat((float) value);
+            else if (cl == Double.TYPE) out.writeDouble((double) value);
+            else if (cl == Short.TYPE) out.writeShort((short) value);
+            else if (cl == Character.TYPE) out.writeChar((char) value);
+            else if (cl == Boolean.TYPE) out.writeBoolean((boolean) value);
+            else throw new InternalError();
+        }
+    }
+
+    private void writeObjectFieldDesc(DataOutputStream out) throws IOException {
+        for (Map.Entry<NameAndType<?>, Object> entry : objectFields.entrySet()) {
+            Class<?> cl = entry.getKey().type();
+            assert !cl.isPrimitive();
+            // obj_typecode
```

Comment:
Try replacing 141-150 with cl.descriptorString(). It looks like it will do all the encoding of the class.